service/api_patch.go

```diff
@@ -35,36 +35,102 @@ type PatchAPIResponse struct {
 	Patch   *patch.Patch `json:"patch"`
 }
 
+// getAuthor returns the author for the patch. If githubAuthor or patchAuthor is provided and exists, will use this instead
+// of the submitter if the submitter is authorized to submit patches on behalf of users.
+// Returns the author, status code, and error.
+func (as *APIServer) getAuthor(data patchData, dbUser *user.DBUser, projectId, patchID string) (string, int, error) {
+	author := dbUser.Id
+	if data.GithubAuthor == "" && data.PatchAuthor == "" {
+		return author, http.StatusOK, nil
+	}
+
+	opts := gimlet.PermissionOpts{
+		Resource:      projectId,
+		ResourceType:  evergreen.ProjectResourceType,
+		Permission:    evergreen.PermissionPatches,
+		RequiredLevel: evergreen.PatchSubmitAdmin.Value,
+	}
+	if !dbUser.HasPermission(opts) {
+		return "", http.StatusUnauthorized, errors.New("user is not authorized to patch on behalf of other users")
+	}
+
+	if data.GithubAuthor != "" {
+		specifiedUser, err := user.FindByGithubName(data.GithubAuthor)
+		if err != nil {
+			return "", http.StatusInternalServerError, errors.Wrap(err, "error looking for specified author")
+		}
+		if specifiedUser != nil {
+			grip.Info(message.Fields{
+				"message":               "overriding patch author as specified by the submitter",
+				"submitter":             dbUser.Id,
+				"new_author":            specifiedUser.Id,
+				"given_github_username": data.GithubAuthor,
+				"patch_id":              patchID,
+			})
+			author = specifiedUser.Id
+		}
+		grip.DebugWhen(specifiedUser == nil, message.Fields{
+			"message":         "github user not found",
+			"github_username": data.GithubAuthor,
+			"patch_id":        patchID,
+		})
+	} else if data.PatchAuthor != "" {
+		specifiedUser, err := user.FindOneById(data.PatchAuthor)
+		if err != nil {
+			return "", http.StatusInternalServerError, errors.Wrap(err, "error looking for specified author")
```

Comment:
It might be a good idea to add the `data.PatchAuthor `to the error message.

Reply:
well this error would be hit if it's a database error, not if the user isn't found, so it doesn't matter too much but I can add it